File: get_golden_cloth_from_yahoo.pl
#!/usr/bin/perl
use strict;
use warnings;

use JSON;
use Time::Piece;
use DateTime;
use POSIX qw(strftime);

my $threshold = 90;
my $today = localtime->strftime('%Y%m%d');

my $mode = $ARGV[0];#P,S,G(プライム、スタンダード、グロース)
if( ! defined $mode || $mode eq ''){ $mode = "P"; };#指定なければプライムだけ

# メインルーチン
my %tikers = %{&load_mst()};
my $len = scalar(keys %tikers);
my $count = 0;
my %result = %{&load_result($today)};

foreach my $t (sort keys %tikers){
    $count++;

    #if($t ne "1941"){ next; }

    if( $mode =~ "P" && $tikers{$t}{class} =~ /^プライム/ ){}
    elsif( $mode =~ "S" && $tikers{$t}{class} =~ /^スタンダード/){}
    elsif( $mode =~ "G" && $tikers{$t}{class} =~ /^グロース/ ){}
    else{ next; }

    print STDERR "$count/$len t:$t $tikers{$t}{name} $tikers{$t}{class}\n";
    my $symbol = "$t.T";
    
    my ($last_end_date, @prices) = get_stock_data($symbol);

    if($#prices < 74){
        print STDERR "  - prices is short. skip. len: $#prices\n";
        next;
    }

    if (grep { !defined($_) } @prices) {
        print STDERR "princes has null. skip\n";
        next;
    }
    
    my @ma25 = moving_average(\@prices, 25);
    my @ma75 = moving_average(\@prices, 75);

    my %r = detect_cross(\@prices, \@ma25, \@ma75, $threshold);
    
    if($r{golden_cross} == 1 || $r{dead_cross} == 1 || $r{golden_cross_near} == 1 || $r{dead_cross_near} == 1){
        $result{$t} = ();
        $result{$t}{golden_cross} = $r{golden_cross};
        $result{$t}{dead_cross} = $r{dead_cross};
        $result{$t}{golden_cross_near} = $r{golden_cross_near};
        $result{$t}{dead_cross_near} = $r{dead_cross_near};
        $result{$t}{price} = $prices[-1];
        $result{$t}{end_date} = strftime("%Y-%m-%dT%H:%M:%S", gmtime($last_end_date));
    }

    #last
}

my $now = DateTime->now(time_zone => 'GMT');
my $formatted_date = $now->strftime('%Y-%m-%dT%H:%M:%S');

my %out = (
    date_modified => $formatted_date,
    result => \%result
    );

print encode_json(\%out);

#&save_json($today, encode_json(\%result));
#&save_json("latest", encode_json(\%result));


#My %result = ($prices[0] => $prices[1]);

#open my $fh, '>', $output_file or die "Could not open file '$output_file': $!";
#print $fh encode_json(\%result);
#close $fh;
#print encode_json(\%result);

exit 0;


################################################################################
### SubRoutines ################################################################
################################################################################

sub load_mst {
    my $tikers_file = "./tickers.json";
    
    open(IN, $tikers_file);
    my $mst_jtxt = <IN>;
    close IN;
    my $mst_json = from_json($mst_jtxt);
    
    return $mst_json;
}

sub load_result {
    my ($date) = @_;
    my $result_file = "./result/$date.json";

    my $json = {};
    $json->{result} = {};
    
    if(-e $result_file){
        open(IN, $result_file);
        my $jtxt = <IN>;
        close IN;
        $json = from_json($jtxt);
    }        
    return $json->{result};
}

sub save_json {
    my ($ticker, $json) = @_;

    my $out = "./result/$ticker.json";
    my $tmpfile = $out."tmp";

    open( OUT, "> $tmpfile" ) or die $!;
    print OUT $json;
    close OUT;
    rename($tmpfile, $out);
    
}

# 株価データの取得
sub get_stock_data {
    my ($symbol) = @_;

    my $end_date = '0';
    my @prices = ();

    eval{
        my $url = '"https://query2.finance.yahoo.com/v8/finance/chart/'.$symbol.'?range=6mo&interval=1d"';
        
        #my $url = "https://www.alphavantage.co/query?function=TIME_SERIES_DAILY&symbol=$symbol&apikey=$api_key";
        #my $response = $ua->get($url);
        my $response = `curl -sS $url`;

        #die "Failed to get data: ", $response->status_line unless $response->is_success;

        #my $data = decode_json($response->decoded_content);
        #print STDERR "ret:$response";
        my $data = decode_json($response);
        #my @prices = reverse( @{$data->{chart}{result}[0]{indicators}{quote}[0]{close}} );
        #splice(@prices, 76);

        @prices = @{$data->{chart}{result}[0]{indicators}{quote}[0]{close}};

        $end_date = $data->{chart}{result}[0]{meta}{currentTradingPeriod}{regular}{end};
        
        #    foreach my $date (sort keys %$time_series) {
        #        push @prices, $time_series->{$date}->{'4. close'};
        #    }
    };
    if($@){
        print STDERR "ERROR:symbol: $symbol\n$@";
    }
    return ($end_date, @prices);
}

# 移動平均の計算
sub moving_average {
    my ($prices, $days) = @_;
    my @ma;
    for my $i (0..$#$prices) {
        if ($i >= $days - 1) {
            my $sum = 0;
            for (my $j = $i - $days + 1; $j <= $i; $j++) {
                my $p = 0;
                if(defined $prices->[$j]){
                    $p = $prices->[$j];
                }
                $sum += $p;
            }
                #$sum += $prices->[$_] for ($i - $days + 1..$i);
            push @ma, $sum / $days;
        } else {
            push @ma, undef;
        }
    }

    #print STDERR "ma$days: @ma\n";
    return @ma;
}

# ゴールデンクロス・デッドクロスの検出
sub detect_cross {
    my ($prices, $ma25, $ma75, $threshold) = @_;
    my %result = (golden_cross => 0, dead_cross => 0, golden_cross_near => 0, dead_cross_near => 0);
    my $total_price = @$prices;
    my $i = $#$prices;
    #for my $i (1..$#$prices) {
    #next unless defined $ma25->[$i] && defined $ma75->[$i];

    my $price_ma_diff = abs($prices->[$i] - $ma75->[$i]);

    if ($ma25->[$i - 1] < $ma75->[$i - 1] && $ma25->[$i] > $ma75->[$i]) {
        $result{golden_cross} = 1;
    } elsif ($ma25->[$i - 1] > $ma75->[$i - 1] && $ma25->[$i] < $ma75->[$i]) {
        $result{dead_cross} = 1;
    }
    if ($price_ma_diff <= $threshold) {
        my $proximity = 100 * (1 - $price_ma_diff / $threshold);
        if ($prices->[$i] > $ma75->[$i]) {
            $result{golden_cross_near} = $proximity;
        } elsif ($prices->[$i] < $ma75->[$i]) {
            $result{dead_cross_near} = $proximity;
        }
    }

    #}
    return %result;
}

File: get_golden_cloth_from_yahoo.py
import yfinance as yf
import json
import datetime
import os
import sys
from decimal import Decimal, ROUND_HALF_UP
import concurrent.futures # ★追加: マルチスレッド用

THRESHOLD = 90

# ヘルパー関数: 正確な四捨五入を行う（RSIスクリプトより流用）
def round_half_up(number, decimals=1):
    """
    数値を指定した桁数で四捨五入(ROUND_HALF_UP)してfloatで返す
    Noneの場合はNoneを返す
    """
    if number is None:
        return None
    try:
        d = Decimal(str(number))
        exp = Decimal("1." + "0" * decimals)
        return float(d.quantize(exp, rounding=ROUND_HALF_UP))
    except:
        return None

def load_mst():
    if not os.path.exists("tickers.json"):
        sys.stderr.write("ERROR: tickers.json not found.\n")
        return {}
    with open("tickers.json", "r", encoding="utf-8") as f:
        return json.load(f)

def load_result(date):
    result_file = f"./result/{date}.json"
    if os.path.exists(result_file):
        with open(result_file, "r", encoding="utf-8") as f:
            return json.load(f).get("result", {})
    return {}

def get_stock_data(symbol):
    try:
        stock = yf.Ticker(symbol)
        # auto_adjust=Falseに変更し、配当落ち調整なしの株価を取得
        hist = stock.history(period="6mo", interval="1d", auto_adjust=False)
        
        if hist.empty:
            return "0", [], []

        # NaNを含む行を削除（Close列基準）
        clean_hist = hist[["Close"]].dropna()
        prices = clean_hist["Close"].tolist()
        
        # 日付リストを作成 (YYYY-MM-DD)
        dates = [d.strftime('%Y-%m-%d') for d in clean_hist.index]
        
        last_end_date = clean_hist.index[-1].strftime("%Y-%m-%dT%H:%M:%S") if len(clean_hist) > 0 else "0"
        
        return last_end_date, prices, dates
    except Exception as e:
        sys.stderr.write(f"ERROR: {symbol} - {e}\n")
        return "0", [], []

def moving_average(prices, days):
    if len(prices) < days:
        return [None] * len(prices)
    return [None] * (days - 1) + [sum(prices[i - days + 1:i + 1]) / days for i in range(days - 1, len(prices))]

def detect_cross(prices, ma25, ma75, threshold):
    # 配列末尾の要素が存在しない(None)場合は計算不可
    if len(prices) < 2 or ma25[-1] is None or ma75[-1] is None:
        return {"golden_cross": 0, "dead_cross": 0, "golden_cross_near": 0, "dead_cross_near": 0}
    
    # 1つ前のデータもNoneチェック
    if ma25[-2] is None or ma75[-2] is None:
        return {"golden_cross": 0, "dead_cross": 0, "golden_cross_near": 0, "dead_cross_near": 0}

    result = {"golden_cross": 0, "dead_cross": 0, "golden_cross_near": 0, "dead_cross_near": 0}
    
    # ゴールデンクロス / デッドクロス判定
    if ma25[-2] < ma75[-2] and ma25[-1] > ma75[-1]:
        result["golden_cross"] = 1
    elif ma25[-2] > ma75[-2] and ma25[-1] < ma75[-1]:
        result["dead_cross"] = 1
    
    # 接近判定
    price_ma_diff = abs(prices[-1] - ma75[-1])
    if price_ma_diff <= threshold:
        proximity = 100 * (1 - price_ma_diff / threshold)
        if prices[-1] > ma75[-1]:
            result["golden_cross_near"] = proximity
        elif prices[-1] < ma75[-1]:
            result["dead_cross_near"] = proximity
            
    return result

# ★追加: 1銘柄分の処理を独立した関数に切り出し
def process_single_ticker(ticker, info, mode, threshold):
    # 対象市場のフィルタリング
    if mode == "P" and "プライム" in info["class"]:
        pass
    elif mode == "S" and "スタンダード" in info["class"]:
        pass
    elif mode == "G" and "グロース" in info["class"]:
        pass
    else:
        return None # 対象外

    symbol = f"{ticker}.T"
    last_end_date, prices, dates = get_stock_data(symbol)
    
    # データ不足チェック (75日平均を出すため75日分必要)
    if len(prices) < 75:
        sys.stderr.write(f"  - prices is short ({len(prices)} < 75). Skipping {symbol}.\n")
        return None
        
    ma25 = moving_average(prices, 25)
    ma75 = moving_average(prices, 75)
    
    cross_result = detect_cross(prices, ma25, ma75, threshold)
    
    history_data = []
    data_len = len(dates)
    start_idx = max(0, data_len - 60)
    
    for i in range(start_idx, data_len):
        history_data.append({
            "d": dates[i],
            "p": round_half_up(prices[i], 1),
            "m25": round_half_up(ma25[i], 2),
            "m75": round_half_up(ma75[i], 2)
        })
        
    return ticker, {
        **cross_result, 
        "price": prices[-1], 
        "end_date": last_end_date,
        "history": history_data
    }

def main(mode="P"):
    today = datetime.datetime.utcnow().strftime('%Y%m%d')
    tickers = load_mst()
    result = load_result(today)
    
    # 処理対象の銘柄だけをリストアップ
    target_items = [
        (t, info) for t, info in tickers.items()
        if (mode == "P" and "プライム" in info["class"]) or
           (mode == "S" and "スタンダード" in info["class"]) or
           (mode == "G" and "グロース" in info["class"])
    ]
    total = len(target_items)
    count = 0

    # ★変更: ThreadPoolExecutorで並行処理を実行する
    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
        # 順番を維持しつつ並行でデータ取得と計算を行う
        results = executor.map(lambda item: process_single_ticker(item[0], item[1], mode, THRESHOLD), target_items)
        
        for res in results:
            count += 1
            if res is not None:
                t, data = res
                result[t] = data
                sys.stderr.write(f"{count}/{total} t:{t} completed\n")
            else:
                sys.stderr.write(f"{count}/{total} skipped\n")
            
    now = datetime.datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S')
    output = {"date_modified": now, "result": result}
    sys.stdout.write(json.dumps(output, ensure_ascii=False, separators=(',', ':')) + "\n")

if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else "P"
    main(mode)
